ext/gvl_tracing_native_extension/gvl_tracing.c

```diff
@@ -49,21 +49,62 @@
   #define UNUSED_ARG
 #endif
 
-static VALUE tracing_start(VALUE _self, VALUE output_path);
-static VALUE tracing_stop(VALUE _self);
-static double timestamp_microseconds(void);
-static void set_native_thread_id(void);
-static void render_event(const char *event_name);
-static void on_thread_event(rb_event_flag_t event, const rb_internal_thread_event_data_t *_unused1, void *_unused2);
-static void on_gc_event(VALUE tpval, void *_unused1);
-static VALUE mark_sleeping(VALUE _self);
+typedef struct {
+  bool current_thread_seen;
+  unsigned int current_thread_serial;
+  uint64_t thread_id;
+  VALUE thread;
+  rb_event_flag_t previous_state; // Used to coalesce similar events
+  bool sleeping; // Used to track when a thread is sleeping
+} thread_local_state;
+
+#ifdef HAVE_RB_INTERNAL_THREAD_SPECIFIC_GET // 3.3+
+
+static int thread_storage_key = 0;
+
+static size_t thread_local_state_memsize(UNUSED_ARG const void *data) {
+    return sizeof(thread_local_state);
+}
+
+static void thread_local_state_mark(void *data) {
+  thread_local_state *state = (thread_local_state *)data;
+  rb_gc_mark(state->thread);
```

Comment:
Yes, even though that is a bit pointless because the only thing holding this object is the thread itself, so it's not really a concern.

The real reason we're calling `rb_gc_mark` here is to that the thread is "pinned". Ruby 2.7 introduced GC compaction, meaning Ruby objects can be moved around, but to avoid breaking historical C extensions, any reference marked with `rb_gc_mark` is considered pinned and won't be moved.

To mark objects without pining them, extensions can use `rb_gc_mark_movable`, but pinning is what we want here, to ensure `(uint64_t)thread` remains consistent.

That said, I haven't double checked, but I suspect `Thread` objects are not movable, but a bit like `RB_GC_GUARD`, this is to be explicit and future proof.

Reply:
Minor: Maybe worth leaving here a

```c
// Marking thread to make sure it stays pinned
```

comment, as this is definitely a good point!